src/js/media-tooltip.ts

```diff
@@ -0,0 +1,278 @@
+  updateXOffset = () => {
+    const position = this.position;
+
+    // we don't offset against tooltips coming out of left and right sides
+    if (position === 'left' || position === 'right') {
+      // could have been offset before switching to a new position
+      this.style.removeProperty('--media-tooltip-offset-x');
+      return;
+    }
+
+    // We need to calculate the difference (diff) between the left edge of the
+    // tooltip compared to the left edge of the container element, to see if it
+    // bleeds out (and the same for the right edges).
+    // If they do, then we apply the diff as an offset to get it back within bounds
+    // + any extra margin specified to create some buffer space, so it looks better.
+    // e.g. it's 20px out of bounds, we nudge it 20px back in + margin
+    const tooltipStyle = getComputedStyle(this);
+    const containingSelector = this.containWithin ?? defaultContainerSelector;
+    const containingEl = closestComposedNode(this, containingSelector);
+    if (!containingEl) return;
+    const { x: containerX, width: containerWidth } =
+      containingEl.getBoundingClientRect();
+    const { x: tooltipX, width: tooltipWidth } = this.getBoundingClientRect();
+    const tooltipRight = tooltipX + tooltipWidth;
+    const containerRight = containerX + containerWidth;
+    const offsetXVal = tooltipStyle.getPropertyValue(
+      '--media-tooltip-offset-x'
+    );
+    const currOffsetX = offsetXVal
+      ? parseFloat(offsetXVal.replace('px', ''))
+      : 0;
+    const marginVal = tooltipStyle.getPropertyValue(
+      '--media-tooltip-container-margin'
+    );
+    const currMargin = marginVal ? parseFloat(marginVal.replace('px', '')) : 0;
+
+    // We might have already offset the tooltip previously so we remove it's
+    // current offset from our calculations, because we need to know if it goes
+    // outside the boundary if we weren't already offsetting it
+    // We also add on any additional container margin specified. Depending on
+    // if we're adjusting the element leftwards or rightwards, we need either a
+    // positive or negative offset
+    const leftDiff = tooltipX - containerX + currOffsetX - currMargin;
+    const rightDiff = tooltipRight - containerRight + currOffsetX + currMargin;
+
+    // out of left bounds
+    if (leftDiff < 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${leftDiff}px`);
+      return;
+    }
+
+    // out of right bounds
+    if (rightDiff > 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${rightDiff}px`);
+      return;
+    }
+
+    // no spilling out
+    this.style.removeProperty('--media-tooltip-offset-x');
+  };
+
+  /**
+   * Get or set tooltip position
+   */
+  get position(): TooltipPosition | undefined {
+    return getStringAttr(this, Attributes.POSITION);
+  }
+
+  set position(value: TooltipPosition | undefined) {
+    setStringAttr(this, Attributes.POSITION, value);
+  }
```

Comment:
what do you think of renaming `position` to `placement` or `place`?

- the reasoning is in CSS the `position` property is something else, there are other props with `place-*` that are more similar to what we're doing here.
- the most popular floating library (floating UI) uses `placement` as well, as does Shoelace and Spectrum web comps

Reply:
> what do you think of renaming `position` to `placement` or `place`?

`placement` seems like a good one for me. Add to the list:
- MUI - https://mui.com/material-ui/react-tooltip/#positioned-tooltips
- Chakra - https://v2.chakra-ui.com/docs/components/tooltip/usage#placement
- Ant Design - https://ant.design/components/tooltip#tooltip-demo-placement

Probably others as well.

Reply:
The one thing in the background that I mentioned to Adam that's (hopefully) upcoming, re: "position" is - https://developer.chrome.com/blog/anchor-positioning-api

Reply:
yep, the anchor API is a good callout. one complex puppy of an API :) but will be powerful stuff

Reply:
Have updated this to placement 👍